### DISCUSSION TITLE
How to prevent a private property from being serialized in upsert

### DISCUSSION THREAD

**Discussion by stefaanv** [mikro-orm/mikro-orm#5324] [Q&A] [answered]
I'm sure there will be a simple solution to this problem, I just haven't been able to find it in the documentation and on the internet
I have this business object :
```ts
class MeteringResume {
  from = new Date()
  till = new Date()
  consumption = 0

  private _lastExceededReport?: Date
  constructor(){
    /...
  }
  //... and some methods
}
```

From which I derive an entity using entitySchema 
```ts
export const MeteringResumeEntity = new EntitySchema<
  Omit<MeteringResume, 'monthPeakExceeding' | 'startQuarterValues' | 'lastExceededReport'>
>({
  name: 'metering',

  properties: {
    from: { type: 'timestamp', primary: true },
    till: { type: 'timestamp' },
    consumption: { type: 'float' },
  },
})
```
when doing an upsert (meterValues is the business object, not the entity), the `_lastExceededReport` private property is added to the SQL query field list which causes an error.
```ts
await em.upsert(MeteringSnapshotEntity, meterValues))
```

My question:  how can I instruct mikro-orm to disregard the private property in the upsert operation ?  Or am I making a mistake when using the business object in the upsert command ?

**Comment by B4nan** [comment]
> Or am I making a mistake when using the business object in the upsert command ?

Yes, probably. Either pass in an entity object, or a POJO - but for POJO its your responsibility to have only valid properties in there.

**Reply by stefaanv** [reply]
This turns out to be a hard fight against the Typescript compiler.
I'm trying to add a `toEntity()`  function to the business object that cleans up the private fields, but typescript complains because the private fields officially don't exist at compile time.  They do exist at runtime unfortunately.
Whilst picking the values renders it own class of messy typescript errors like
```
 Type '() => EntityName<{ from: Date; till: Date; consumption: number; injection: number; batCharge: number; batDischarge: number; gas: number; tariff: Tariff; monthPeakValue: number; monthPeakTime: Date; }>' is not assignable to type '() => EntityName<Omit<MeteringResume, "toEntity" | "startQuarterValues" | "monthPeakExceeding">>'.
```

**Reply by B4nan** [reply]
I'll be happy to help, but I have no idea what you are doing, you'll need to provide a repro first.

(but it sounds like you are doing that "my business entity is not ORM entity and vice versa" approach, which I always discourage in the first place :] you only get a false sense of separation while you are still tightly coupled to the ORM, and you just add more obstacles to using it)

**Reply by stefaanv** [reply]
I guess that is exaclty the trap I fell into.  Where can I read up on how to avoid the `BO vs ORM entity` problem ?
I do like the `EntitySchema` syntax a lot better that the `reflect-metadata` approach though.  I'd like to preserve that if possible.
Also, the entity has behavior which is kind of essential.

I made my repo public, it's at https://github.com/stefaanv/hass-energy-pricing
The culprit is in `src/metering/metering.service.ts` at line 114 where the `this.resume` object is `upserted` to the database
```ts
await em.upsert(MeteringResumeEntity, this.resume.toEntity())
```
I've managed to make it work by implementing `toEntity()` and `fromEntity()` functions in the `MeteringResume` but I had a hard time to make the Typescript compiler accept my solution.
Basically, the `MeteringResume` class has private variables that do not exist in the related entity `MeteringResumeEntity`.  The `toEntity()` has to get rid of these or I get SQL `unknown column name` errors.  The pick solution 
```ts
  toEntity() {
    return pick(this, [
      'from',
      'till',
      'consumption',
      'injection',
      'batCharge',
      'batDischarge',
      'gas',
      'tariff',
      'monthPeakValue',
      'monthPeakTime',
    ]) as MeteringResume
```
is the only one that I can make to work because Typescript doesn't allow for `omit`ting the private properties (because they don't exist for Typescript) even though the `omit` solution is DRY-er because it has no need for enumerating the `MeteringResumeEntity` property keys which is plain un-DRY right as it is now.

It would be nice if the `upsert()` function would accept the `MeteringResume` directly but for that I need some way to tell `MeteringResumeEntity` to ignore the private properties when building the SQL upsert instruction.

I tried to make an alternative implementation for the `toEntity()` function in commit 49d4792 (commented out just below the working implementation), but I bump into strange Typescript errors.

**Reply by B4nan** [reply, marked as answer]
> I made my repo public, it's at [stefaanv/hass-energy-pricing](https://github.com/stefaanv/hass-energy-pricing)

That's a full-blown app, not really a repro, too complex for me to understand what's happening just by looking.

> Where can I read up on how to avoid the BO vs ORM entity problem ?

I don't have any resources, if you ask me, those two are the same (and then there is no problem). What you are trying to do means working with the ORM as if it was a query builder, ignoring all its features connected to the UoW basically. You lose all change tracking, and at the same time you'll be fighting how the ORM works internally (as updates require state unless you use QB directly).

Some people approach this by keeping the state and using `em.getReference()` which gives you back the entity instance from context, that's what I would try if I really wanted to go down this path (but personally I wouldn't bother). So this means that when you load entity, it stays in the identity map, then you convert it to your BO, and either you keep a reference to the ORM entity on it, or you use `em.getReference()` to get it back when you want to convert BO to ORM entity. Note that this requires you to keep one context (one ORM fork - per request/scope - not the global one), I see you have soma manual forking in place so that might not be compatible with this approach (but haven't checked too deeply).

Trying to go with `em.upsert` seems a bit weird to me, but it depends on what you are building (and what you expect from your code).

**Reply by stefaanv** [reply]
Is it possible to add methods to the entity using `EntitySchema` ?

**Reply by B4nan** [reply]
Sure thing, you define your own class and the ORM will use that, it can have methods as usual. I am not aware of a single thing that you couldn't do with EntitySchema and would require decorators.

**Reply by stefaanv** [reply]
OK, I'm starting to see the light.
Indeed, instead of using the upsert, I can simply update the entity now.
Works like a charm.
Thanks !
